__app__/crop/ingress_hyper.py

```diff
@@ -0,0 +1,211 @@
+    and returns it as a dict, keyed by Aranet id.
+
+    Arguments:
+        api_key: api key for authentication
+        dt_from: date range from
+        dt_to: date range to
+    Return:
+        success: whether data request was succesful
+        error: error message
+        data_df_dict: dictionary of {aranet_pro_id (str): DataFrame of sensor_data}
+    """
+
+    success = True
+    error = ""
+    data_df_dict = {}
+
+    headers = {
+        "Content-Type": "application/json",
+        "Authorization": "Bearer " + api_key,
+    }
+
+    dt_from_iso = dt_from.strftime("%Y-%m-%dT%H:%M:%S") + "Z"
+    dt_to_iso = dt_to.strftime("%Y-%m-%dT%H:%M:%S") + "Z"
+
+    url = CONST_CHECK_URL_PATH
+
+    params = {
+        "start_time": dt_from_iso,
+        "end_time": dt_to_iso,
+        "metrics": "aranet_temperature,aranet_humidity",
+        "resolution": "10m",
+        "metadata": "true",
+    }
+    response = requests.get(url, headers=headers, params=params, verify=False)
+
+    if response.status_code != 200:
+        error = "Request's [%s] status code: %d" % (url[:70], response.status_code)
+        success = False
+        return success, error, data_df_dict
+    # if we got to here, we have an API response for all T/RH sensors
+    data = response.json()
+    # the metadata contains a mapping between the Hyper id (MAC address) and the
+    # aranet_pro_id
+    device_mapping = data["metadata"]["devices"]
+    timestamps = data["labels"]
+    value_dicts = data["series"]
+    for k, v in device_mapping.items():
+        hyper_id = k
+        aranet_pro_id = v["vendor_device_id"]
+        # make a dataframe from data for this sensor
+        sensor_data = {
+            value_dict["metric"]: value_dict["values"]
+            for value_dict in value_dicts
+            if value_dict["device_id"] == hyper_id
+        }
+        data_dict = {"Timestamp": pd.to_datetime(timestamps)}
+        data_dict.update(sensor_data)
+        data_df = pd.DataFrame(data_dict)
+        for col_name in data_df.columns:
+            if "temperature" in col_name:
+                data_df.rename(columns={col_name: "Temperature"}, inplace=True)
+            elif "humidity" in col_name:
+                data_df.rename(columns={col_name: "Humidity"}, inplace=True)
+        data_df.set_index("Timestamp", inplace=True)
+        if data_df.index.tz is None:
+            data_df.index = data_df.index.tz_localize("UTC")
+        # put this DataFrame into a dict, keyed by the aranet id
+        data_df_dict[aranet_pro_id] = data_df
+
+    return success, error, data_df_dict
+
+
+def _get_sensor_id(aranet_pro_id, engine):
+    session = session_open(engine)
+    # Get the sensor ID as CROP has it
+    query = session.query(
+        SensorClass.id,
+    ).filter(SensorClass.aranet_pro_id == aranet_pro_id)
+    sensor_id = session.execute(query).fetchone()
+    if isinstance(sensor_id, Iterable):
```

Comment:
very neat - I'm never sure whether/why the result of `fetchone()` is an iterable.

Reply:
Yeah, this caught me by surprise here too. Not sure what the logic is.